patches/minecraft/net/minecraft/world/World.java.patch

```diff
@@ -17,7 +17,18 @@
      private int field_181546_a = 63;
      protected boolean field_72999_e;
      public final List<Entity> field_72996_f = Lists.<Entity>newArrayList();
-@@ -107,6 +114,12 @@
+     protected final List<Entity> field_72997_g = Lists.<Entity>newArrayList();
+-    public final List<TileEntity> field_147482_g = Lists.<TileEntity>newArrayList();
+-    public final List<TileEntity> field_175730_i = Lists.<TileEntity>newArrayList();
++    public final net.minecraftforge.common.util.ChunkedTileEntityList field_147482_g = new net.minecraftforge.common.util.ChunkedTileEntityList();
++    public final net.minecraftforge.common.util.ChunkedTileEntityList field_175730_i = new net.minecraftforge.common.util.ChunkedTileEntityList();
+     private final List<TileEntity> field_147484_a = Lists.<TileEntity>newArrayList();
+     private final List<TileEntity> field_147483_b = Lists.<TileEntity>newArrayList();
++    public List<net.minecraft.util.math.ChunkPos> tileEntitiesChunkToBeRemoved = Lists.<net.minecraft.util.math.ChunkPos>newArrayList();
```

Comment:
adding and iterations arent the main problem, its mostly the removing part. Although some alternatives you suggested may or may not improve on the first two operations, they may influence removing in a more negative way (which is the main problem atm).